cpp/src/fil/fil.cu

```diff
@@ -543,12 +547,12 @@ void tree2fil_dense(std::vector<dense_node>* pnodes, int root,
 }
 
 template <typename fil_node_t, typename T, typename L>
-int tree2fil_sparse(std::vector<fil_node_t>* pnodes, const tl::Tree<T, L>& tree,
+int tree2fil_sparse(std::vector<fil_node_t>& nodes, const int root,
```

Comment:
Instead of using a non-const reference as a parameter, it's better to:
- use a pointer parameter (as was used before);
- inside the function, dereference the pointer to create a local reference variable.

Reply:
Instead of a separate vector of nodes and root, consider passing an iterator parameter that points to where the root node will be added. If necessary, feel free to add an extra template parameter for that.

This is up to you, however.

Reply:
I'll take a closer look at passing an iterator. I have a vague memory that I considered and rejected this for some specific reason, but that may have been somewhere else in the code.

As for your first recommendation, that would put us at odds with C++ core guidelines, which recommend passing a raw pointer only when the function specifically manipulates the address (taking advantage of the "pointeriness" of the parameter) and passing smart pointers only to take advantage of data ownership semantics. Otherwise, references are recommended. It would also put us at odds [with the Google style guide](https://google.github.io/styleguide/cppguide.html#Inputs_and_Outputs), require additional error checking for null pointers, and preclude the possibility of taking advantage of reference collapsing rules. See additional discussion on this point as it relates to cpplint [here](https://github.com/cpplint/cpplint/issues/148) and arguments against the recommendation to use pointers as output parameters in the old (now corrected) version of the GSG on the basis of C++ semantic intent [here](https://eyakubovich.github.io/2018-11-27-google-cpp-style-guide-is-no-good/).

I haven't corrected this elsewhere in the conversion code, but since I was already modifying the function signature, I figured I'd fix it now.

Reply:
Thanks for the clarification. I have no objections to using the reference here.

Reply:
Great, thanks! Now that I've looked at it again, I'm remembering that I avoided passing the iterator on readability grounds. I implemented both and decided that it would be slightly clearer to future generations if everything was referenced by index. It also doesn't really fit the pattern of how I would expect a function to behave if it took in an iterator, but I don't have a specific guideline to point to for that. Since it seems like neither of us feel super strongly about it, I'll leave it as is and mark it as resolved, but we can hash it out further if you're worried about it.